core/src/main/scala/kafka/server/metadata/ZkMetadataCache.scala

```diff
@@ -248,7 +258,12 @@ class ZkMetadataCache(brokerId: Int, metadataVersion: MetadataVersion, brokerFea
   }
 
   override def getAliveBrokerNode(brokerId: Int, listenerName: ListenerName): Option[Node] = {
-    metadataSnapshot.aliveBrokers.get(brokerId).flatMap(_.getNode(listenerName))
+    val snapshot = metadataSnapshot
+    brokerId match {
+      case id if snapshot.controllerId.filter(_.isInstanceOf[KRaftCachedControllerId]).exists(_.id == id) =>
+        kraftControllerNodeMap.get(id)
+      case _ => snapshot.aliveBrokers.get(brokerId).flatMap(_.getNode(listenerName))
+    }
```

Comment:
We should be able to use pattern matching a little more cleanly like:

```suggestion
    val snapshot = metadataSnapshot
    snapshot.controllerId match {
      case Some(controllerId: KRaftCachedControllerId) => kraftControllerNodeMap.get(controllerId.id)
      case _ => snapshot.aliveBrokers.get(brokerId).flatMap(_.getNode(listenerName))
    }
```

Reply:
This is wrong. In Kraft brokers and Zk brokers with KRaft controller. It always returns controller information for any broker node query.

Reply:
Ah, i see, i missed `exists` part. This could still be cleaned up with pattern matching, but it's not critical for this PR